alica_engine/src/engine/teammanager/TeamManager.cpp

```diff
@@ -1,92 +1,146 @@
+    if (_useAutoDiscovery) {
+        _agentAnnouncementTimeInterval = AlicaTime::seconds(sc["Alica"]->get<unsigned long>("Alica.AgentAnnouncementTimeInterval", NULL));
+        _announcementRetries = sc["Alica"]->get<int>("Alica.AnnouncementRetries", NULL);
     }
+    readSelfFromConfig();
+    constructAnnouncementMessage();
 }
 
-TeamManager::~TeamManager()
+void TeamManager::constructAnnouncementMessage()
 {
-    for (auto& agentEntry : _agents) {
-        delete agentEntry.second;
+    _presenceMessage.senderID = _localAgent->getId();
+    _presenceMessage.token = _localAgent->getToken();
+    _presenceMessage.senderName = _localAgent->getName();
+    _presenceMessage.senderSdk = _localAgent->getSdk();
+    _presenceMessage.planHash = _localAgent->getPlanHash();
+    _presenceMessage.role = _localAgent->getProperties().getDefaultRole();
+    for (const auto& cap : _localAgent->getProperties().getCharacteristics()) {
+        const std::string& name = cap.first;
+        const std::string& value = cap.second->getCapValue()->getName();
+        _presenceMessage.capabilities.push_back(std::make_pair(name, value));
     }
 }
 
+TeamManager::~TeamManager() {}
+
 void TeamManager::setTeamTimeout(AlicaTime t)
 {
+    AgentsCache::AgentMap& agents = *_agentsCache.get();
     _teamTimeOut = t;
-    for (auto& agentEntry : _agents) {
+
+    for (auto& agentEntry : agents) {
         agentEntry.second->setTimeout(t);
     }
 }
 
-void TeamManager::readTeamFromConfig()
+void TeamManager::readSelfFromConfig()
 {
     essentials::SystemConfig& sc = essentials::SystemConfig::getInstance();
-    std::string localAgentName = _engine->getRobotName();
-    std::shared_ptr<std::vector<std::string>> agentNames = sc["Globals"]->getSections("Globals.Team", NULL);
-    bool foundSelf = false;
-    for (const std::string& agentName : *agentNames) {
-        int id = sc["Globals"]->tryGet<int>(-1, "Globals", "Team", agentName.c_str(), "ID", NULL);
-
-        Agent* agent = new Agent(_engine, _teamTimeOut, _engine->getId(id), agentName);
-        if (!foundSelf && agentName == localAgentName) {
-            foundSelf = true;
-            _localAgent = agent;
-            _localAgent->setLocal(true);
-        } else {
-            for (auto& agentEntry : _agents) {
-                if (*(agentEntry.first) == *(agent->getId())) {
-                    AlicaEngine::abort("TM: Two robots with the same ID in Globals.conf. ID: ", agent->getId());
-                }
-            }
-        }
-        _agents.emplace(agent->getId(), agent);
-    }
-    if (!foundSelf) {
-        AlicaEngine::abort("TM: Could not find own agent name in Globals Id = ", localAgentName);
+    const std::string localAgentName = _engine->getRobotName();
+    int id = sc["Local"]->tryGet<int>(-1, "Local", localAgentName.c_str(), "ID", NULL);
+
+    AgentAnnouncement aa;
+    if (id != -1) {
+        aa.senderID = _engine->getId(id);
+    } else {
+        aa.senderID = _engine->generateId(8);
```

Comment:
If you want to hardcode the size, at least make it a constexpr and give it a name.